module/bdev/lvol/vbdev_lvol_rpc.c

```diff
@@ -1505,3 +1508,269 @@ rpc_bdev_lvol_shallow_copy_status(struct spdk_jsonrpc_request *request,
+	if (bdev == NULL) {
+		SPDK_ERRLOG("bdev '%s' does not exist\n", req.name);
+		spdk_jsonrpc_send_error_response(request, -ENODEV, spdk_strerror(ENODEV));
+		goto cleanup;
+	}
+
+	lvol = vbdev_lvol_get_from_bdev(bdev);
+	if (lvol == NULL) {
+		SPDK_ERRLOG("lvol does not exist\n");
+		spdk_jsonrpc_send_error_response(request, -ENODEV, spdk_strerror(ENODEV));
+		goto cleanup;
+	}
+
+	// Create a bitmap recording the allocated clusters
+	cluster_size = spdk_bs_get_cluster_size(lvol->lvol_store->blobstore);
+	block_size = spdk_bdev_get_block_size(bdev);
+	num_blocks = spdk_bdev_get_num_blocks(bdev);
+	lvol_size = num_blocks * block_size;
+
+	if (req.offset + req.size > lvol_size) {
+		SPDK_ERRLOG("offset %lu and size %lu exceed lvol size %lu\n",
+			    req.offset, req.size, lvol_size);
+		spdk_jsonrpc_send_error_response_fmt(request, -EINVAL,
+						     "offset %lu and size %lu exceed lvol size %lu",
+						     req.offset, req.size, lvol_size);
+		goto cleanup;
+	}
+
+	segment_size = req.size;
+	if (req.size == 0) {
+		segment_size = lvol_size;
+	}
+
+	if (!spdk_is_divisible_by(req.offset, cluster_size) ||
+	    !spdk_is_divisible_by(segment_size, cluster_size)) {
+		SPDK_ERRLOG("offset %lu and size %lu must be a multiple of cluster size %lu\n",
+			    req.offset, segment_size, cluster_size);
+		spdk_jsonrpc_send_error_response_fmt(request, -EINVAL,
+						     "offset %lu and size %lu must be a multiple of cluster size %lu",
+						     req.offset, segment_size, cluster_size);
+		goto cleanup;
+	}
+
+	num_clusters = spdk_divide_round_up(segment_size, cluster_size);
+	fragmap = spdk_bit_array_create(num_clusters);
+	if (fragmap == NULL) {
+		SPDK_ERRLOG("failed to allocate fragmap with num_clusters %lu\n", num_clusters);
+		spdk_jsonrpc_send_error_response(request, -ENOMEM,
+						 spdk_strerror(ENOMEM));
+		goto cleanup;
+	}
+
+	// Construct a fragmap of the lvol
+	rc = spdk_bdev_open_ext(bdev->name, false,
+				dummy_bdev_event_cb, NULL, &desc);
+	if (rc != 0) {
+		spdk_jsonrpc_send_error_response(request, rc, spdk_strerror(-rc));
+		goto cleanup;
+	}
+
+	channel = spdk_bdev_get_io_channel(desc);
+	if (channel == NULL) {
+		spdk_bdev_close(desc);
+		SPDK_ERRLOG("could not allocate I/O channel.\n");
+		spdk_jsonrpc_send_error_response(request, -ENOMEM,
+						 spdk_strerror(ENOMEM));
+		goto cleanup;
+	}
+
+	io = spdk_zmalloc(sizeof(struct fragmap_io), 0, NULL,
+			  SPDK_ENV_LCORE_ID_ANY, SPDK_MALLOC_DMA);
+	if (io == NULL) {
+		spdk_put_io_channel(channel);
+		spdk_bdev_close(desc);
+		spdk_jsonrpc_send_error_response(request, -ENOMEM,
+						 spdk_strerror(ENOMEM));
+		goto cleanup;
+	}
+
+	io->bdev = bdev;
+	io->bdev_desc = desc;
+	io->bdev_io_channel = channel;
+	io->request = request;
+	io->fragmap = fragmap;
+	io->num_clusters = num_clusters;
+	io->block_size = block_size;
+	io->cluster_size = cluster_size;
+	io->offset = req.offset;
+	io->size = segment_size;
+	io->current_offset = req.offset;
+
+	rc = spdk_bdev_seek_data(desc, channel,
```

Comment:
Here we have to check rc

Reply:
Yeah, reclaimed the resources. Thank you.